### PR TITLE
Activity Log: Add the notion of retention limits + an upsell for them

### PR DESCRIPTION
**Pull request by mattgawarecki**
Resolves `1200412004370260-as-1200532973279678`.

#### Changes proposed in this Pull Request

* Create a new feature flag `activity-log/retention-policies` to enable/disable Activity Log retention policies (disabled by default for now).
* Create a new set of reducers, actions, etc. for storing and retrieving a site's Activity Log retention policy.
* Create a new component `RetentionLimitUpsell` to offer customers the ability to upgrade their Activity Log retention policy, visible when they reach the oldest Activity they're allowed to access.
* Add analytics events for when people see this new upsell, and for when they click its call-to-action.
* Modify `ActivityCardList` to only show events within the bounds of the current site's retention policy, if the feature flag is enabled.
* Modify the Activity Log to reflect a loading state until the current site's retention policy is known.

#### Known issues

* The button/link in the call-to-action currently redirects to `/pricing`, because a context-appropriate upsell page doesn't yet exist.
* If an error occurs while fetching the site's retention policy, the Activity Log will show a loading state forever.
* This implementation only covers Calypso Green. We will presumably need a similar pull request to cover Calypso Blue, borrowing functionality seen here.

#### Testing instructions

**IMPORTANT:** To test this pull request, you must configure your testing environment so that it accesses a development sandbox version of the WordPress.com API, and that sandbox must have `D64586-code` applied to it. Otherwise, you'll only see a loading state on the Activity Log page after the request to get your site's retention policy fails.

##### Setup

1. Before starting Calypso Green, change the value of the `activity-log/retention-policies` feature flag to true for your environment; or, add `?flags=activity-log/retention-policies` to your browser's URL as you test.
2. Select a site with events in its Activity Log, and visit its Activity Log.

##### Upsell visible

1. If your site has more than 30 days of events, scroll that far back in the Activity Log. Verify you see an upsell prompt on the last page, as shown in the reference screenshots.
2. Manually change your site's retention policy in the application state by running (e.g.) `state.activityLog.retentionPolicy[ <your_site_id> ].days = 1;` in the development console. Click on the filter bar at the top of the page, then dismiss it to cause the list of Activities to re-render.
3. Verify the list displays appropriately, hiding activities that would fall outside the retention period you set.
4. Verify that the upsell is visible and references the number of days you set in step 2.

##### Upsell not visible: flag disabled

1. Load the Activity Log for a site where you would normally see the retention limit upsell (see test cases above).
2. Add `?flags=-activity-log/retention-policies` to your browser's URL and hit Enter/Return to reload the page (note the minus sign).
3. Verify the retention limit upsell is no longer visible, and activities that would fall outside the retention policy are no longer hidden.

##### Upsell not visible: not enough activities

1. Load the Activity Log for a site with 30 days of events or fewer; alternatively, increase the retention period by modifying the application state as described in the "Upsell visible" case above.
2. Verify that you do not see the retention limit upsell described in the previous case, even scrolling to the last available page.

##### Analytics / tracking

1. Load the Activity Log for a site where you would normally be able to see the retention limit upsell (see test cases above), but _do not_ scroll it into view yet.
2. Open the Network panel of your browser's development console so you can see outgoing requests to `t.gif`.
3. Watching the Network panel as you do so, scroll the upsell into view.
4. Verify that only once the "Upgrade storage" button is fully in view, a request is sent with the event name `calypso_activitylog_retentionlimit_view`. A `site_id` parameter should be also included in the request.
5. Scroll back up the page until the upsell is no longer in view, then scroll back down. Verify the request isn't sent a second time.
6. Click the "Upgrade storage" button, verifying that another request is sent with the event name `calypso_activitylog_retentionlimit_upgrade_click` and a `site_id` parameter.

#### Reference screenshots

Screen Shot 2021-07-26 at 10 52 52 Screen Shot 2021-07-26 at 10 52 31

### CODE REVIEW

File: client/components/activity-card-list/retention-limit-upsell/index.tsx [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+					) }
+				</p>
+				<Button
+					primary
+					ref={ upsellRef }
+					className="retention-limit-upsell__call-to-action-button"
+					onClick={ trackUpgradeClick }
+					href="/pricing"
+				>
+					{ translate( 'Upgrade storage' ) }
```
Comment (a8ci18n) [inline client/components/activity-card-list/retention-limit-upsell/index.tsx:96]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **27** times:
`translate( 'Storage Upgrade!' )` *ES Score: 9*

File: client/components/activity-card-list/index.jsx [hunk 2]
```diff
@@ -21,14 +23,24 @@ import ActivityCard from 'calypso/components/activity-card';
...
 import QueryRewindCapabilities from 'calypso/components/data/query-rewind-capabilities';
 import QueryRewindState from 'calypso/components/data/query-rewind-state';
+import RetentionLimitUpsell from 'calypso/components/activity-card-list/retention-limit-upsell';
 
 /**
  * Style dependencies
  */
 import './style.scss';
 
+const getRetentionPolicyRequestStatus = ( state, siteId ) => {
```
Comment (mattgawarecki) [inline client/components/activity-card-list/index.jsx:26]: This may warrant its own file eventually, but I wasn't sure if that time would be now or later.

**Review by Luchadores — APPROVED** [review]
From my side this is looking good, Matt. 👍

File: client/state/action-types.ts [hunk 3]
```diff
@@ -33,6 +33,12 @@ export const ACTIVE_PROMOTIONS_RECEIVE = 'ACTIVE_PROMOTIONS_RECEIVE';
 export const ACTIVE_PROMOTIONS_REQUEST = 'ACTIVE_PROMOTIONS_REQUEST';
 export const ACTIVE_PROMOTIONS_REQUEST_FAILURE = 'ACTIVE_PROMOTIONS_REQUEST_FAILURE';
 export const ACTIVE_PROMOTIONS_REQUEST_SUCCESS = 'ACTIVE_PROMOTIONS_REQUEST_SUCCESS';
+export const ACTIVITY_LOG_RETENTION_POLICY_REQUEST = 'ACTIVITY_LOG_RETENTION_POLICY_REQUEST';
```
Comment (cbauerman) [inline client/state/action-types.ts:36]: Did you consider using a newer pattern like React Query ( p4TIVU-9AF-p2 )? The redux state pattern is perfectly fine, imo, but I think the newer patterns involve much less boilerplate.

File: client/components/activity-card-list/index.jsx (hunk 2 above)
Comment (cbauerman) [reply client/components/activity-card-list/index.jsx:26]: if it is only used here I think YAGNI applies 👍

File: client/components/activity-card-list/retention-limit-upsell/index.tsx [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+							}
+						)
+					) }
+				</p>
+				<Button
+					primary
+					ref={ upsellRef }
+					className="retention-limit-upsell__call-to-action-button"
+					onClick={ trackUpgradeClick }
+					href="/pricing"
```
Comment (cbauerman) [inline client/components/activity-card-list/retention-limit-upsell/index.tsx]: this should be `/pricing/:site`

File: client/components/activity-card-list/retention-limit-upsell/style.scss [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+		height: 80px;
+		padding: initial;
+	}
+
+	// Don't allow any interaction with this element;
+	// it's only for display purposes
+	user-select: none;
+	pointer-events: none;
+
+	mask-image: linear-gradient(
```
Comment (cbauerman) [inline client/components/activity-card-list/retention-limit-upsell/style.scss:12]: This is very cool, and looks great in practice!

**Review by cbauerman — APPROVED** [review]
Looks good! Nothing I commented on is blocking, so LGTM 🚢

File: client/state/action-types.ts (hunk 3 above)
Comment (mattgawarecki) [reply client/state/action-types.ts:36]: To be completely honest, I wasn't aware React Query had been blessed for use in Calypso yet. 😅 I'll keep that in mind for future PRs, though!

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/6341727

Thank you @mattgawarecki for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/6341727).

**Pull request merged**